### ISSUE TITLE
Bug: @typescript-eslint node module is 280mb on disk

### ISSUE THREAD

**Issue by gkiely** [typescript-eslint/typescript-eslint#5464]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

https://github.com/gkiely/vite-react-starter

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Packages use typescript peer dependency. typescript-eslint node_module is 11mb.

### Actual Result

Packages use typescript, typescript-eslint node_module is 280mb.

### Versions

typescript-eslint@latest

**Comment by JoshuaKGoldberg** [comment]
tl;dr: this is happening in your repro repo, but not on a fresh install. Weird.

<details>
<summary>Showing this doesn't happen on a fresh install</summary>

For a fresh install in a new package, I get 87M on disk with npm. Reproduction technique (long console text trimmed):

```plaintext
$ rm -rf ./*
zsh: no matches found: ./*
$ npm init --yes
Wrote to /Users/josh/repos/tseslinttemp/package.json: ...
$ npm install --save-dev @typescript-eslint/parser @typescript-eslint/eslint-plugin eslint typescript
added 130 packages, and audited 131 packages in 11s
$ du -sh ./node_modules/
 87M	./node_modules/
```

Looking at all the folders 100k or higher:

```plaintext
$ du -sh ./node_modules/* | sort -h

 64M	./node_modules/typescript
7.6M	./node_modules/@typescript-eslint
3.7M	./node_modules/eslint
1.1M	./node_modules/ajv
1.0M	./node_modules/esquery
700K	./node_modules/@eslint
632K	./node_modules/tsutils
584K	./node_modules/uri-js
480K	./node_modules/acorn
476K	./node_modules/js-yaml
400K	./node_modules/eslint-utils
308K	./node_modules/regexpp
276K	./node_modules/fast-glob
260K	./node_modules/@nodelib
252K	./node_modules/semver
248K	./node_modules/grapheme-splitter
224K	./node_modules/type-fest
184K	./node_modules/argparse
176K	./node_modules/@humanwhocodes
124K	./node_modules/doctrine
116K	./node_modules/flatted
112K	./node_modules/picomatch
100K	./node_modules/eslint-scope
...
```

So, 64M (73.5%) of that size comes from `typescript` alone. `@typescript-eslint/*` itself only takes up 7.6M (8.7%) of that.

</details>

<details>
<summary>Aside: looking at some odd dist/lib/*.d.ts outputs...</summary>

```plaintext
$ du -sh ./node_modules/@typescript-eslint/* | sort -h    
 32K	./node_modules/@typescript-eslint/experimental-utils
 68K	./node_modules/@typescript-eslint/visitor-keys
208K	./node_modules/@typescript-eslint/types
1.2M	./node_modules/@typescript-eslint/utils
2.3M	./node_modules/@typescript-eslint/scope-manager
 66M	./node_modules/@typescript-eslint/typescript-estree
 69M	./node_modules/@typescript-eslint/parser
 70M	./node_modules/@typescript-eslint/type-utils
 73M	./node_modules/@typescript-eslint/eslint-plugin
```

Digging in:

```plaintext
$ du -sh ./node_modules/@typescript-eslint/scope-manager/dist/* | sort -h
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ID.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ID.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ID.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ID.js.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ScopeManager.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/ScopeManager.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/analyze.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/analyze.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/analyze.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/analyze.js.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/assert.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/assert.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/assert.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/assert.js.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/index.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/index.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/index.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/index.js.map
8.0K	./node_modules/@typescript-eslint/scope-manager/dist/ScopeManager.js.map
 12K	./node_modules/@typescript-eslint/scope-manager/dist/ScopeManager.js
 80K	./node_modules/@typescript-eslint/scope-manager/dist/variable
240K	./node_modules/@typescript-eslint/scope-manager/dist/definition
240K	./node_modules/@typescript-eslint/scope-manager/dist/referencer
376K	./node_modules/@typescript-eslint/scope-manager/dist/scope
1.3M	./node_modules/@typescript-eslint/scope-manager/dist/lib
```

...why is `scope-manager/dist/lib` so big?

```plaintext
josh ~/repos/vite-react-starter $ du -sh ./node_modules/@typescript-eslint/scope-manager/dist/lib/* | sort -h
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/base-config.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/base-config.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/base-config.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/base-config.js.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.iterable.d.ts
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.iterable.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.iterable.js
...(dozens more)
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/webworker.iterable.d.ts.map
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/webworker.iterable.js
4.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/webworker.iterable.js.map
8.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/es5.js
8.0K	./node_modules/@typescript-eslint/scope-manager/dist/lib/index.js
 12K	./node_modules/@typescript-eslint/scope-manager/dist/lib/index.d.ts
 12K	./node_modules/@typescript-eslint/scope-manager/dist/lib/webworker.js.map
 20K	./node_modules/@typescript-eslint/scope-manager/dist/lib/webworker.js
 32K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.js.map
 60K	./node_modules/@typescript-eslint/scope-manager/dist/lib/dom.js
```

So, that's kind of weird. I can file a separate bug on it: either it's unintentional or it's a quirk that should be documented somewhere.

</details>

Responding to both this issue and https://github.com/typescript-eslint/typescript-eslint/issues/3622#issuecomment-1212247732:

> I'll just note typescript is installed 4 times in this node_module when using npm. `eslint-plugin`, `parser`, `type-utils`, `typescript-estree` all contain `typescript`. This contributes to the package being 280mb on disk.

My default instinct would be to say that if all your packages are using the latest version, they should resolve to just one version of TypeScript.

For anybody else who is seeing duplicate versions of `typescript` or other essential packages:
* If you're on npm, `npm dedupe` might help
* If you're on Yarn v1 (not the fancy new Berry version), `npx yarn-deduplicate` might help

I tried `npm dedupe` on the `vite-react-starter` repo and it didn't help, though.

```plaintext
$ du -sh ./node_modules/@typescript-eslint/eslint-plugin/node_modules/* | sort -h
 52K	./node_modules/@typescript-eslint/eslint-plugin/node_modules/estraverse
 72K	./node_modules/@typescript-eslint/eslint-plugin/node_modules/tslib
100K	./node_modules/@typescript-eslint/eslint-plugin/node_modules/eslint-scope
252K	./node_modules/@typescript-eslint/eslint-plugin/node_modules/semver
632K	./node_modules/@typescript-eslint/eslint-plugin/node_modules/tsutils
4.1M	./node_modules/@typescript-eslint/eslint-plugin/node_modules/@typescript-eslint
 64M	./node_modules/@typescript-eslint/eslint-plugin/node_modules/typescript
```

Why is this duplicate TypeScript there? 🤔

**Comment by JoshuaKGoldberg** [comment]
Ah! The nested `typescript` package is `4.7.4`.

```plaintext
$ cat package.json | grep typescript\":
    "typescript": "^4.8.0-beta",
$ cat ./node_modules/typescript/package.json | grep version
    "version": "4.8.0-dev.20220809",
$ cat ./node_modules/@typescript-eslint/eslint-plugin/node_modules/typescript/package.json | grep version
    "version": "4.7.4",
```

`eslint-plugin` declares TypeScript as an optional peer dependency: https://github.com/typescript-eslint/typescript-eslint/blob/d0239104f4dbffd2b5ecdb19e520c7d4b71962e0/packages/eslint-plugin/package.json#L73-L77

Deleting `package-lock.json` and `node_modules/` in the repo and running again doesn't change the installed versions. `package-lock.json` still has _four_ entries you can find by searching for `/typescript": {`:

```json
"node_modules/@typescript-eslint/eslint-plugin/node_modules/typescript": {
  "version": "4.7.4",
  "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.7.4.tgz",
  "integrity": "sha512-C0WQT0gezHuw6AdY1M2jxUO83Rjf0HP7Sk1DtXj6j1EwkQNZrHAg2XPWlq62oqEhYvONq5pkC2Y9oPljWToLmQ==",
  "dev": true,
  "peer": true,
  "bin": {
    "tsc": "bin/tsc",
    "tsserver": "bin/tsserver"
  },
  "engines": {
    "node": ">=4.2.0"
  }
},
```

```jsonc
"node_modules/@typescript-eslint/type-utils/node_modules/typescript": {
  "version": "4.7.4",
  "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.7.4.tgz",
  "integrity": "sha512-C0WQT0gezHuw6AdY1M2jxUO83Rjf0HP7Sk1DtXj6j1EwkQNZrHAg2XPWlq62oqEhYvONq5pkC2Y9oPljWToLmQ==",
  // (same)
},
```

```jsonc
"node_modules/@typescript-eslint/typescript-estree/node_modules/typescript": {
  "version": "4.7.4",
  "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.7.4.tgz",
  "integrity": "sha512-C0WQT0gezHuw6AdY1M2jxUO83Rjf0HP7Sk1DtXj6j1EwkQNZrHAg2XPWlq62oqEhYvONq5pkC2Y9oPljWToLmQ==",
  // (same)
},
```

```json
"node_modules/typescript": {
  "version": "4.8.0-dev.20220809",
  "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.8.0-dev.20220809.tgz",
  "integrity": "sha512-MfDvCrf/GqKM/GwLuhs2ykgauNUg2TCR5YzIDj+b2scDwRE9XCRcgQAWh29omMmGFIAhP91QXKt+aXV0hchkxw==",
  // (same)
},
```

...so, why isn't npm deduplicating the versions? Is it because the `typescript` package is at a beta version?

Switching the root `package.json` to have `typescript` at version `"^4.7.4"` instead of `"^4.8.0-beta"` **does fix the issue**. `package-lock.json` now has just one entry under `typescript`.

I think this is an npm issue (bug? quirk? intentional design decision?) external to our package. What do you think @gkiely?

**Comment by bradzacher** [comment]
First up - when you use npm to install it automatically will install peer dependencies of all packages. This feature was added in a recent version of npm and was widely regarded as a bad idea, but sadly it's now here to stay in the `npm` cli. I don't believe other package managers have this behaviour.

Second - you have chosen to install the TS 4.8 beta.

The version `4.8.0-beta` is a pre-release version which does not match the dependency range `"*"` (which is the range we use).

Putting these together - when npm installs it attempts to install our peer dependency on TS.

However it sees that your root version of TS doesn't match, and thus it needs to install the latest compatible version of TS (I.e. `4.7.4`).

However because you already have a root version of the package, npm cannot install this new version at the root. So the only othe pla d to install it is as a dependency of the modules that require it.
So it adds a copy of it to each of the packages that need it.

Thus you end up with a bunch of copies of TS installed.

If you used a compatible, non pre-release TS version - you won't see this behaviour because npm can hoist the version to the root. Which is why @JoshuaKGoldberg isn't seeing this in a fresh install.

**Comment by JoshuaKGoldberg** [comment]
@bradzacher I think we should add an entry to https://typescript-eslint.io/docs/linting/troubleshooting on this one. I wouldn't have expected it, and have in the past gotten teams to use TypeScript nightlies. Thoughts?

IIRC someone (you?) had mentioned that `*` doesn't match beta versions, which is very surprising IMO.

**Comment by bradzacher** [comment]
`*` does match beta versions in some contexts.

I don't think this worth adding noise to the FAQ because this isn't an FAQ. I doubt we'll have someone run into this again:

- most users don't do fresh installs that often
- most users don't use a pre-release version of TS as their main TS version.
- TS pre-release versions are relatively short-lived.

You can try it out but you would probably find that if you do your install with the "latest" TS version and then upgrade the version to the beta, there will only be one version of TS installed.
You'll also probably find that yarn1 behaves differently here (which would be different again to yarn2 and pnpm).
